### DISCUSSION TITLE
issue: watch does not trigger useEffect when you watch an array field but trigger if you watch all fields

### DISCUSSION THREAD

**Discussion by patrickn2** [react-hook-form/react-hook-form#11391] [Q&A]
### Version Number

7.49.2

### Codesandbox/Expo snack

https://codesandbox.io/p/sandbox/watch-array-fields-zsltd5

### Steps to reproduce

1. Click on add Spec value
2. The Spec Value counter will not increase
3. Replace the useEffect dependency at line 79 for a watch that watch all fields "allFields"
4. Click on add Spec Value
5. The Spec Value counter will increase normally

### Expected behaviour

The Spec Value counter should increase when you watch an array field

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
> watch result is optimised for render phase instead of useEffect's deps, to detect value update you may want to use an external custom hook for value comparison.

https://react-hook-form.com/docs/useform/watch

**Comment by patrickn2** [comment]
Hey @bluebill1049 Thank you for your answer.

Do you have an example of the use of this external hook?
I just need to check if a field inside the array is changed to recalculate another field and watching all fields makes my form kinda slow.

**Comment by chanthinh** [comment]
Hi @patrickn2, Did you solve this issue?

**Comment by patrickn2** [comment]
> Hi @patrickn2, Did you solve this issue?

Unfortunately no I did not.
I'm still watching all form fields and this sucks :(

**Comment by chanthinh** [comment]
@patrickn2,
Could you check this one: 
https://codesandbox.io/p/sandbox/watch-array-fields-forked-pw4wf8?file=%2Fsrc%2FApp.tsx%3A83%2C17
I don't know if that's what you expected or not

**Comment by patrickn2** [comment]
> @patrickn2, Could you check this one: https://codesandbox.io/p/sandbox/watch-array-fields-forked-pw4wf8?file=%2Fsrc%2FApp.tsx%3A83%2C17 I don't know if that's what you expected or not

After number 8 it stops counting.

**Comment by chanthinh** [comment]
@patrickn2,
Try this one: https://codesandbox.io/p/sandbox/watch-array-fields-forked-pw4wf8?file=%2Fsrc%2FApp.tsx%3A66%2C59

**Comment by patrickn2** [comment]
Thank you Chan
Gonna try to use useMemo instead of useEffect.
This does not solve the watch issue but it is an option.

**Comment by seunggs** [comment]
Hi @patrickn2, did you find a solution for this? This still seems to be an issue. Looks like it might be useEffect not being triggered due to array mutation like this issue: https://stackoverflow.com/questions/63536876/react-hooks-useeffect-not-being-triggered-on-array-of-objects
